File: geekbang/algorithm-models/6.2-并查集代码模板/DisjointSet.java
// Java
class DisjointSet {
    public DisjointSet(int n) {
        fa = new int[n];
        for (int i = 0; i < n; i++) fa[i] = i;
    }


    public int find(int x) {
        if (x == fa[x]) return x;
        return fa[x] = find(fa[x]);
    }


    public void unionSet(int x, int y) {
        x = find(x);
        y = find(y);
        if (x != y) fa[x] = y;
    }


    int[] fa;
};

File: geekbang/algorithm-models/6.2-并查集代码模板/DisjointSet.cpp
//C/C++
class DisjointSet {
public:
    DisjointSet(int n) {
        fa = vector<int>(n, 0);
        for (int i = 0; i < n; i++) fa[i] = i;
    }


    int find(int x) {
        if (x == fa[x]) return x;
        return fa[x] = find(fa[x]);
    }


    void unionSet(int x, int y) {
        x = find(x), y = find(y);
        if (x != y) fa[x] = y;
    }


private:
    vector<int> fa;
};
